### PR TITLE
Add Array::with_nulls

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/6528

# Rationale for this change
 
Propose adding a new safe API to replace nulls

@tustvold  mentioned the potential implications of this API https://github.com/apache/arrow-rs/issues/6528#issuecomment-2401522424 here. I added an example to show what was going on

# What changes are included in this PR?

Add `Array::with_nulls`

# Are there any user-facing changes?

New API

### CODE REVIEW

File: arrow-array/src/array/mod.rs [hunk 1]
```diff
@@ -196,6 +198,28 @@ pub trait Array: std::fmt::Debug + Send + Sync {
...
     fn nulls(&self) -> Option<&NullBuffer>;
 
+    /// Replaces the nulls of this array.
+    ///
+    /// # Panics
+    /// Panics if the length of the null buffer is not equal to the length of the array.
+    ///
+    /// # Example:
+    /// ```
+    /// # use arrow_array::{Array, Int32Array};
```
Comment (alamb) [inline arrow-array/src/array/mod.rs:208]: Here is the new API and an example. I am not sure this is a good idea due to @tustvold 's comment here: https://github.com/apache/arrow-rs/issues/6528#issuecomment-2401522424 but figured I would make a PR for discussion (and only now do I really understand what he is talking about)

File: arrow-array/src/array/mod.rs [hunk 2]
```diff
@@ -196,6 +198,28 @@ pub trait Array: std::fmt::Debug + Send + Sync {
...
+    /// # use arrow_array::cast::AsArray;
+    /// # use arrow_buffer::NullBuffer;
+    /// // Create an array with values [1, null, 3, 4, 5]
+    /// let array = Int32Array::from(vec![Some(1), None, Some(3), Some(4), Some(5)]);
+    /// // Set the first, third, and fifth elements to null, others to valid
+    /// let nulls = Some(NullBuffer::from(vec![false, true, false, true, false]));
+    /// let array_with_nulls = array.with_nulls(nulls);
+    /// assert_eq!(
+    ///   array_with_nulls.as_primitive(),
+    ///   &Int32Array::from(vec![None, Some(0), None, Some(4), None]),
```
Comment (alamb) [inline arrow-array/src/array/mod.rs:218]: the fact that this is now `Some(0)` (whatever value was created for the `null` element in the original array) could have potential issues 🤔

**Review by tustvold — CHANGES_REQUESTED** [review]
As written this is unsound, in particular it allows unmasking indexes within dictionaries that aren't valid.

I think the only reasonable behaviour is to combine the null masks, at which point this is just a special case of the nullif kernel.

**Comment by alamb** [comment]
I agree -- let's not do this. I have another idea that might help https://github.com/apache/arrow-rs/issues/6528#issuecomment-2449860007

**Pull request closed without merging**